Add b2b_ecommerce app to handle bulk enrollment code purchases

noisecapella opened the pull request:
#### Pre-Flight checklist

- [x] Testing
  - [x] Code is tested

#### What are the relevant tickets?
Part of #858

#### What's this PR do?
Implements the REST API and backend functionality to allow customers to purchase enrollment codes

#### How should this be manually tested?
No great way to do that yet, probably best to wait until the front end PR is in place

noisecapella commented:
In addition to everything else, whoever reviews this let me know what you think of the naming and organization of everything.

gsidebo commented on ecommerce/api.py:
```diff
@@ -915,3 +925,85 @@ def get_or_create_data_consents(basket):
     return data_consents
+
+
+def create_coupons(
```
This method has 14 parameters. Seems like a major code smell to me. I don't have any bright ideas off the top of my head for refactoring it, but worth mentioning

noisecapella replied:
I agree, I talked with @rhysyngsun about this earlier. He suggested making sure the parameters were explicit so they could be linted, but we didn't have any other ideas to improve this.

noisecapella commented on b2b_ecommerce/models.py:
```diff
@@ -0,0 +1,109 @@
+            return "B2BReceipt with no attached order"
+
+
+class B2BOrderCouponPayment(TimestampedModel):
```
I'm going to remove this and make `coupon_payment` a field on `B2BOrder`

noisecapella commented on b2b_ecommerce/api.py:
```diff
@@ -0,0 +1,130 @@
+        :254
+    ]
+    line_items["item_0_quantity"] = order.num_seats
+    line_items["item_0_sku"] = "enrollment_code"
```
That's a good point. I think it's important for `code` to be for an enrollment code purchase but we could put some information in the `sku` which has the id and type being purchased.

noisecapella commented on b2b_ecommerce/api.py:
```diff
@@ -0,0 +1,130 @@
+from mitxpro.utils import now_in_utc
+
+
+_REFERENCE_NUMBER_PREFIX = "XPRO-ENROLLMENT-"
```
Any suggestions @pdpinch?

noisecapella commented on b2b_ecommerce/urls.py:
```diff
@@ -0,0 +1,17 @@
+urlpatterns = [
+    url("^api/b2b_checkout/$", B2BCheckoutView.as_view(), name="b2b-checkout"),
+    url(
+        "^api/b2b_order_fulfillment/$",
```
Sounds good, I'll make the change

noisecapella commented:
@gsidebo I also added a `unique_id` field to `B2BOrder` in my commit. It will be an autogenerated UUID which can be used for lookup in the receipt page.

noisecapella commented on b2b_ecommerce/api.py:
```diff
@@ -0,0 +1,130 @@
+_REFERENCE_NUMBER_PREFIX = "XPRO-ENROLLMENT-"
+
+
+def make_b2b_reference_id(order):
```
I made this a property function on the two models

noisecapella commented on b2b_ecommerce/api.py:
```diff
@@ -0,0 +1,130 @@
+    line_items["item_0_tax_amount"] = "0"
+    line_items["item_0_unit_price"] = str(price)
+
+    payload = {
```
I'm not sure how to do that. It's not just the price and two URLs, the `B2BOrder` doesn't have `Line` for example. I did some refactoring to try to simplify this logic, let me know if you still see areas which can be consolidated.

noisecapella commented:
Ok, finally addressed all feedback. FYI I deleted `B2BOrderCouponPayment` because I realized this is a one-to-many relation and not a many-to-many so a foreign key would suffice.

gsidebo commented on b2b_ecommerce/models.py:
```diff
@@ -0,0 +1,102 @@
+    objects = OrderManager(REFERENCE_NUMBER_PREFIX)
+
+    @property
+    def reference_id(self):
```
This is identical in both models. Could this just be moved to `OrderAbstract`? I suggested implementing the prefix as a model class property in another comment. If you do that I think it would work

gsidebo commented on b2b_ecommerce/constants.py:
```diff
@@ -0,0 +1,2 @@
+"""Constants for b2b_ecommerce"""
+REFERENCE_NUMBER_PREFIX = "XPRO-ENROLLMENT-"
```
Still think it wouldn't hurt to change this to be a little less arbitrary. `MITXPRO-B2B-` maybe

gsidebo commented on b2b_ecommerce/api.py:
```diff
@@ -0,0 +1,92 @@
+            coupon_type=CouponPaymentVersion.SINGLE_USE,
+        )
+        order.coupon_payment_version = payment_version
+        order.save()
```
`save_and_log`?

noisecapella replied:
I wanted to hold off on `save_and_log` until the order fulfillment happens in the view functions so that there is only one record of change instead of multiple records in the audit table. But I do need to move `save_and_log` in `B2BCheckoutView` to be after this function so that it captures the output correctly.

gsidebo commented on b2b_ecommerce/views.py:
```diff
@@ -0,0 +1,137 @@
+        receipt.order = order
+        receipt.save()
+
+        decision = request.data["decision"]
```
Take or leave: This is another method that is almost identical to the non-bulk version. To reduce the risk of that, maybe this block from 109-128 can be moved into a separate API method called `determine_order_status(order, decision)` et. al. which would return the status string

noisecapella replied:
I tried to go in this direction but it's tricky because of the `return Response(status=200)`

gsidebo replied:
Seems like it's functionally equivalent if you return `None` in that function under those conditions, then in the `post` function just `return Response(status=200)` if the status is `None`. The function could be called `determine_order_status_change` to make it clearer. If there's no status change, there's no need to do anything else with this order, so you can just return from there.

noisecapella replied:
Ah I see. I just made the change

noisecapella commented:
@gsidebo I made the change and it should be ready for review again
